python/knext/knext/operator/op.py

```diff
@@ -0,0 +1,194 @@
+from knext.operator.base import BaseOp
+from knext.operator.invoke_result import InvokeResult
+from knext.operator.spg_record import SPGRecord
+
+cache = knext.common.cache.LinkCache(500, 60)
+
+
+class ExtractOp(BaseOp, ABC):
+    """Base class for all knowledge extract operators."""
+
+    def __init__(self, params: Dict[str, str] = None):
+        super().__init__(params)
+
+    def invoke(self, record: Dict[str, str]) -> List[Dict[str, str]]:
+        raise NotImplementedError(
+            f"{self.__class__.__name__} need to implement `invoke` method."
+        )
+
+    @staticmethod
+    def _pre_process(*inputs):
+        return (inputs[0],)
+
+    @staticmethod
+    def _post_process(output) -> Dict[str, Any]:
+        if isinstance(output, InvokeResult):
+            return output.to_dict()
+        if isinstance(output, tuple):
+            return InvokeResult[List[SPGRecord]](*output[:3]).to_dict()
+        else:
+            return InvokeResult[List[SPGRecord]](output).to_dict()
+
+
+class LinkOp(BaseOp, ABC):
+    """Base class for all entity link operators."""
+
+    bind_to: SPGTypeName
+
+    bind_schemas: Dict[SPGTypeName, str] = {}
+
+    def __init__(self):
+        super().__init__()
+
+    def invoke(self, property: str, subject_record: SPGRecord) -> List[SPGRecord]:
+        raise NotImplementedError(
+            f"{self.__class__.__name__} need to implement `invoke` method."
+        )
+
+    def _handle(self, *inputs) -> Dict[str, Any]:
+        _property, subject_record = self._pre_process(*inputs)
+        cache_key = str(self.bind_to) + _property
+        cache_property = cache.get(cache_key)
+        if cache_property:
+            output = [SPGRecord(spg_type_name=self.bind_to).upsert_property("id", cache_property)]
+        else:
+            output = self.invoke(_property, subject_record)
+        post_output = self._post_process(output)
+        return post_output
+
+    @staticmethod
+    def _pre_process(*inputs):
+        return inputs[0], SPGRecord.from_dict(inputs[1])
+
+    @staticmethod
+    def _post_process(output) -> Dict[str, Any]:
+        if isinstance(output, InvokeResult):
+            return output.to_dict()
+        if isinstance(output, tuple):
+            return InvokeResult[List[SPGRecord]](*output[:3]).to_dict()
+        else:
+            return InvokeResult[List[SPGRecord]](output).to_dict()
+
+
+class FuseOp(BaseOp, ABC):
+    """
+    Base class for all entity fuse operators.
+    """
+
+    """"""
+    bind_to: SPGTypeName
+
+    bind_schemas: Dict[SPGTypeName, str] = {}
+
+    def __init__(self):
+        super().__init__()
+
+    def link(self, subject_record: SPGRecord) -> List[SPGRecord]:
+        raise NotImplementedError(
+            f"{self.__class__.__name__} need to implement `link` method."
+        )
+
+    def merge(
+        self, subject_record: SPGRecord, linked_records: List[SPGRecord]
+    ) -> List[SPGRecord]:
```

Comment:
Why is the output of the `FuseOp` a `List[SPGRecord]`?  It should be to merge the current `SPGRecord` with an existing `SPGRecord`, right?